pulsar-client/src/main/java/org/apache/pulsar/client/api/ClientBuilder.java

```diff
@@ -0,0 +1,263 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pulsar.client.api;
+
+import java.io.Serializable;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+
+import org.apache.pulsar.client.api.PulsarClientException.UnsupportedAuthenticationException;
+
+/**
+ * Builder interface that is used to construct a {@link PulsarClient} instance.
+ */
+public interface ClientBuilder extends Serializable, Cloneable {
+
+    /**
+     * @return the new {@link PulsarClient} instance
+     */
+    PulsarClient build() throws PulsarClientException;
+
+    /**
+     * Create a copy of the current client builder.
+     * <p>
+     * Cloning the builder can be used to share an incomplete configuration and specialize it multiple times. For
+     * example:
+     *
+     * <pre>
+     * ClientBuilder builder = PulsarClient.builder().ioThreads(8).listenerThreads(4);
+     *
+     * PulsarClient client1 = builder.clone().serviceUrl(URL_1).build();
+     * PulsarClient client2 = builder.clone().serviceUrl(URL_2).build();
+     * </pre>
+     */
+    ClientBuilder clone();
```

Comment:
nit: `@Override`?

Reply:
`clone()` is a bit weird in the way it's treated. It cannot be marked as `@Override`